src/ocf_request.c

```diff
@@ -97,14 +97,9 @@ struct ocf_request *ocf_req_new(ocf_queue_t queue, ocf_core_t core,
 	if (unlikely(!req))
 		return NULL;
 
-	if (map_allocated) {
-		req->map = req->__map;
-		req->alock_status = (uint8_t*)&req->__map[core_line_count];
-		req->alloc_core_line_count = core_line_count;
-	} else {
-		req->alloc_core_line_count = 1;
-	}
-
+	req->map = req->__map;
+	req->alloc_core_line_count = map_allocated ? core_line_count : 1;
+	req->alock_status = (uint8_t*)&req->__map[req->alloc_core_line_count];
```

Comment:
In standard I/O path whenever we fail to allocate map here, there is an attempt to allocate it from outside the memory pool (ocf_req_alloc_map called from ocf_core_volume_submit_io). After this change, ocf_req_alloc_map() will allways think that the map is allocated. This will result in out of bound access in case of large read/write I/O. It can easily be reproduced with a test similar to the one you added, just replace this one line
    core.submit_flush(0, core_device.size.bytes)
with
Rio().readwrite(ReadWrite.WRITE).bs(Size(core_device.size.bytes)).size(Size(core_device.size.bytes)).target(core).run()

Reply:
I don't see any elegant way to fix this. I guess maybe flush should have its own map allocation logic, just like discard does (ocf_req_alloc_map_discard). This newly added ocf_req_alloc_map_flush would be called from ocf_core_volume_submit_flush and it would set req->map = req->__map in case req->map == NULL. This is good enough for flush (as it doesn't really need the map) and works with the case of large R/W I/O.

Reply:
I don't understand what you wrote regrading the failed allocation. In lines 91-94 I saw that if a "core_line_count" allocation from the pool fails then 1 core_line_count is allocated from the pool and not from outside of it.

Reply:
Yes, in ocf_req_new() after failed allocation from the pool we attempt to alloc only 1 core line fro the pool. If this succeeds (the map_allocated = false case in line 100)  req->map remains NULL. Then in case of regular I/O (not discard) this results in ocf_req_alloc_map() calling env_zalloc() in an attempt to alloc multiple core lines separaterly from the request. 

On your branch the case that I describe above will have req->map != NULL. This will not trigger env_zalloc() allocation in ocf_req_alloc_map(). Further request processing will attempt to dereference req->map at indexes above 1, outside allocation bounds.

Are my instructions for the test clear to you? If so, you should be able to reproduce this easily in pyocf.

Reply:
Thanks for the extra information. I'll try to reproduce with your test case and see if I can propose a better solution.